pkg/activator/net/revision_backends.go

```diff
@@ -224,11 +225,9 @@ func (rw *revisionWatcher) checkDests(dests sets.String) {
 	if len(dests) == 0 {
 		// We must have scaled down.
 		rw.clusterIPHealthy = false
-
 		rw.logger.Debug("ClusterIP is no longer healthy.")
-
-		// Send update that we are now inactive (both params invalid).
-		rw.sendUpdate("", nil)
+		// Send update that we are now inactive (Deleted true).
+		rw.sendUpdate("", nil, true)
```

Comment:
Is this really a case for "Deleted: true"? AFAICT this is not necessarily called on deleted revisions?

Reply:
I see and stopped calling it here now.

Current code's behavior calls `sendUpdate()` when endpoint is removed by `endpointsUpdate()`. So, the log message prints `Failed to get revision throttler for revision` twice as

<details>
<summary>activaotr log when ksvc is deleted</summary>
{"level":"debug","ts":"2019-09-17T12:05:42.382Z","logger":"activator","caller":"activator/main.go:324","msg":"No signal yet.","commit":"b066e90","knative.dev/controller":"activator"}
{"level":"debug","ts":"2019-09-17T12:05:45.211Z","logger":"activator","caller":"net/throttler.go:371","msg":"Revision delete \"default/hello-example-qfmnx\"","commit":"b066e90","knative.dev/controller":"activator"}
{"level":"debug","ts":"2019-09-17T12:05:45.327Z","logger":"activator","caller":"net/revision_backends.go:397","msg":"Endpoints updated: &v1.Endpoints{TypeMeta:v1.TypeMeta{Kind:\"\", APIVersion:\"\"}, ObjectMeta:v1.ObjectMeta{Name:\"hello-example-qfmnx-rxlfk\", GenerateName:\"\", Namespace:\"default\", SelfLink:\"/api/v1/namespaces/default/endpoints/hello-example-qfmnx-rxlfk\", UID:\"701f1263-d943-11e9-a46f-02674d032d6c\", ResourceVersion:\"1949069\", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63704318727, loc:(*time.Location)(0x295b500)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string{\"app\":\"hello-example-qfmnx\", \"networking.internal.knative.dev/serverlessservice\":\"hello-example-qfmnx\", \"networking.internal.knative.dev/serviceType\":\"Private\", \"serving.knative.dev/configuration\":\"hello-example\", \"serving.knative.dev/configurationGeneration\":\"1\", \"serving.knative.dev/revision\":\"hello-example-qfmnx\", \"serving.knative.dev/revisionUID\":\"6f89ddd6-d943-11e9-a46f-02674d032d6c\", \"serving.knative.dev/service\":\"hello-example\"}, Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Initializers:(*v1.Initializers)(nil), Finalizers:[]string(nil), ClusterName:\"\"}, Subsets:[]v1.EndpointSubset(nil)}","commit":"b066e90","knative.dev/controller":"activator"}
{"level":"debug","ts":"2019-09-17T12:05:45.327Z","logger":"activator","caller":"net/revision_backends.go:408","msg":"Updating Endpoints: \"default/hello-example-qfmnx\" (backends: 0)","commit":"b066e90","knative.dev/controller":"activator"}
{"level":"debug","ts":"2019-09-17T12:05:45.327Z","logger":"activator","caller":"net/revision_backends.go:228","msg":"ClusterIP is no longer healthy.","commit":"b066e90","knative.dev/controller":"activator"}
{"level":"error","ts":"2019-09-17T12:05:45.327Z","logger":"activator","caller":"net/throttler.go:380","msg":"Failed to get revision throttler for revision \"default/hello-example-qfmnx\"","commit":"b066e90","knative.dev/controller":"activator","error":"revision.serving.knative.dev \"hello-example-qfmnx\" not found","stacktrace":"knative.dev/serving/pkg/activator/net.(*Throttler).handleUpdate\n\t/home/knakayam/.go/src/knative.dev/serving/pkg/activator/net/throttler.go:380\nknative.dev/serving/pkg/activator/net.(*Throttler).Run\n\t/home/knakayam/.go/src/knative.dev/serving/pkg/activator/net/throttler.go:316"}
{"level":"debug","ts":"2019-09-17T12:05:45.403Z","logger":"activator","caller":"net/revision_backends.go:426","msg":"Deleting endpoint \"default/hello-example-qfmnx\"","commit":"b066e90","knative.dev/controller":"activator"}
{"level":"error","ts":"2019-09-17T12:05:45.403Z","logger":"activator","caller":"net/throttler.go:380","msg":"Failed to get revision throttler for revision \"default/hello-example-qfmnx\"","commit":"b066e90","knative.dev/controller":"activator","error":"revision.serving.knative.dev \"hello-example-qfmnx\" not found","stacktrace":"knative.dev/serving/pkg/activator/net.(*Throttler).handleUpdate\n\t/home/knakayam/.go/src/knative.dev/serving/pkg/activator/net/throttler.go:380\nknative.dev/serving/pkg/activator/net.(*Throttler).Run\n\t/home/knakayam/.go/src/knative.dev/serving/pkg/activator/net/throttler.go:316"}
{"level":"debug","ts":"2019-09-17T12:05:59.916Z","logger":"activator","caller":"activator/main.go:324","msg":"No signal yet.","commit":"b066e90","knative.dev/controller":"activator"}
</details>

Although I added `sendUpdate()` to follow current code flow, it should be sent from endpointsDeleted() later.

Reply:
Hmm, I don't know if that's true to be honest. We can reach this piece of code even if we don't delete the revision I think. What will happen then?